### DISCUSSION TITLE
How to increase the accuracy of model from chapter 04. PyTorch Custom Datasets

### DISCUSSION THREAD

**Discussion by nabeelalikhan0** [mrdbourke/pytorch-deep-learning#1317] [Q&A]
I tried many regularization techniques to fix this but...
Epoch: 50 | Train Loss: 0.5506 | Train Acc: 0.8003 | Test Loss: 1.6564 | Test Acc: 0.1875
Total training time: 44.556 seconds
that's the max i got 

`
class Model_1(nn.Module):
    def __init__(self,input_shape:int,hidden_units:int,output_shape:int):
        super().__init__()
        self.block_1 = nn.Sequential(
            nn.Conv2d(input_shape,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.MaxPool2d(2,2)
        )

        self.block_2 = nn.Sequential(
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.MaxPool2d(2,2)
        )
        
        self.block_3 = nn.Sequential(
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Conv2d(hidden_units,hidden_units,3,1,1),
            nn.BatchNorm2d(hidden_units),
            nn.ReLU(),
            nn.Dropout(0.25),
            nn.MaxPool2d(2,2)
        )

        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.LazyLinear(output_shape)
        )

    def forward(self,x):
        x = self.block_1(x)
        x = self.block_2(x)
        x = self.block_3(x)
        x = self.classifier(x)
        return x
    
torch.manual_seed(42)
model = Model_1(3,10,len(classes_names)).to(device)
model

`

And this is my dataset 
`
(Dataset ImageFolder
     Number of datapoints: 349
     Root location: data\images\train
     StandardTransform
 Transform: Compose(
                Resize(size=(96, 96), interpolation=bilinear, max_size=None, antialias=True)
                TrivialAugmentWide(num_magnitude_bins=31, interpolation=InterpolationMode.NEAREST, fill=None)
                ToTensor()
            ),
 Dataset ImageFolder
     Number of datapoints: 100
     Root location: data\images\test
     StandardTransform
 Transform: Compose(
                Resize(size=(96, 96), interpolation=bilinear, max_size=None, antialias=True)
                ToTensor()
            ))
`

Thankyou!

**Comment by mrdbourke** [comment]
Hi @nabeelalikhan0 ,

What does your dataset look like? Can you share some example images?

Have you tried increasing the size of the network? (e.g. adding more layers)

If that doesn't work you could try to use transfer learning and see if that helps: https://www.learnpytorch.io/06_pytorch_transfer_learning/

**Comment by Haji-Suleman** [comment]
Hi @nabeelalikhan0 

From your results:

Train Acc: 0.8003  
Test Acc: 0.1875


This looks like overfitting.

Your model is learning the training data well (80%) but failing on the test set (18%).

Why this is happening
1. Very small dataset

You only have:

349 training images

100 test images

That is quite small for a CNN with many convolution layers.

2. Model is too complex

Your network has:

3 convolution blocks

Multiple Conv layers

BatchNorm everywhere

This is a relatively deep model for such a small dataset, so it likely memorizes instead of generalizing.

What you can try
Option 1: Use transfer learning (recommended)

Instead of training from scratch, use a pretrained model like ResNet18 and fine-tune it on your dataset. This usually works much better for small datasets.

Option 2: Reduce model size

Use fewer Conv layers

Reduce hidden_units

Add more Dropout

Option 3: Check your dataset

Make sure classes are balanced

Visualize some images

Verify labels are correct

**Comment by akashcodes23** [comment]
Classic overfitting! Your model memorized training data. Solutions:

Add Dropout — nn.Dropout(p=0.2) after hidden layers
Add L2 regularization — weight_decay=0.01 in your optimizer
Reduce model complexity — fewer layers/neurons
Get more data or use data augmentation (torchvision.transforms)
Use early stopping — stop when val loss starts increasing

Start with Dropout — it's the fastest fix.
